### PR TITLE
Enhancement/externalize error messages properties

### PR DESCRIPTION
**Pull request by ap891843**
This PR would be able to support the following:

- External properties file for managing messages in the application.

- Support localization.
- Message Locale is fetched from the client-side and is defaulted to English. In case the user picks a locale that is not supported, 
   we display an error notification to the user and switch to the previous valid locale.

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

### CODE REVIEW

**Comment by ap891843** [comment]
@asatklichov 
@Nurkambay 
@temanbrcom

File: server/src/main/resources/message/messages_en.properties [hunk 1]
```diff
@@ -0,0 +1,28 @@
+CobolLineReaderImpl.longLineMsg=Source text cannot go past column 80
```
Comment (temanbrcom) [inline server/src/main/resources/message/messages_en.properties:1]: This key pattern will work for now, but I would prefer storing by purpose, e.g. error/notification, not by class. We may review and change it later.

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * returns currently set locale.
+ */
+@Slf4j
+@Singleton
+public final class LocaleStoreImpl implements LocaleStore {
+  private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+  private final Lock readLock = readWriteLock.readLock();
+  private final Lock writeLock = readWriteLock.writeLock();
+  private final Locale defaultLocale = new Locale("en", "en");
+  private Communications communication;
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java:48]: communications

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+   *
+   * @return {@link LocaleEnum}
+   */
+  @Override
+  public LocaleEnum getSupportedLocale() {
+    readLock.lock();
+    try {
+      if (Objects.isNull(supportedLocale))
+        return Enum.valueOf(LocaleEnum.class, defaultLocale.getCountry().toUpperCase());
+      return supportedLocale;
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java:68]: This "return" should be moved to end of method. Regarding to finally would be more readable

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+   * @param locale to be updated.
+   */
+  @Override
+  public void updateLocale(String locale) {
+    try {
+      LocaleEnum localeEnum = LocaleEnum.valueOf(locale.toUpperCase());
+      setSupportedLocale(localeEnum);
+      Locale updatedLocale = new Locale(locale);
+      notifyList.forEach(consumer -> consumer.accept(updatedLocale));
+    } catch (IllegalArgumentException e) {
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java:95]: Why you catch unchecked (runtime) exception?

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  /**
+   * Update the locale for an application.
+   *
+   * @param locale to be updated.
+   */
+  @Override
+  public void updateLocale(LocaleEnum locale) {
+    try {
+      setSupportedLocale(locale);
+    } catch (IllegalArgumentException e) {
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java:109]: same

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java [hunk 6]
```diff
@@ -66,12 +66,12 @@
     implements TextDocumentService, EventObserver<RunAnalysisEvent> {
   private static final List<String> COBOL_IDS = Arrays.asList("cobol", "cbl", "cob");
   private static final String GIT_FS_URI = "gitfs:/";
-  private static final String GITFS_URI_NOT_SUPPORTED = "GITFS URI not supported";
+  public static final String DESC_ERROR_MSG = "CobolTextDocumentService.descErrorMsg";
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java:69]: This one is also for logs and doesn't need to localize

File: server/src/main/java/com/broadcom/lsp/cobol/service/delegates/communications/ServerCommunications.java [hunk 7]
```diff
@@ -115,7 +118,21 @@ public void notifyThatDocumentAnalysed(String uri) {
...
+  /**
+   * show a supplied message to the client with the supplied {@link MessageType}
+   *
+   * @param messageType {@link MessageType}
+   * @param message to be displayed at client end.
+   */
+  @Override
+  public void notifyGeneralMessage(MessageType messageType, String message) {
+    runAsync(() -> showMessage(messageType, message));
   }
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/delegates/communications/ServerCommunications.java:136]: Where do you call this method?

**Review by temanbrcom — APPROVED** [review]

File: server/src/main/java/com/broadcom/lsp/cobol/service/delegates/communications/ServerCommunications.java (hunk 7 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/delegates/communications/ServerCommunications.java:136]: https://github.com/ap891843/che-che4z-lsp-for-cobol/blob/dd2e6e91b467b1321b59399fa98cc98f4dc05587/server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java#L118

from the localeStore, when the user selects a locale not supported yet.

File: server/src/test/java/com/broadcom/lsp/cobol/TestModule.java [hunk 8]
```diff
@@ -60,6 +61,9 @@ protected void configure() {
     bind(String.class)
         .annotatedWith(named(PATH_TO_TEST_RESOURCES))
         .toProvider(() -> ofNullable(getProperty(PATH_TO_TEST_RESOURCES)).orElse(""));
+    bind(LocaleStore.class).to(LocaleStoreImpl.class);
+    bind(MessageService.class).to(PropertiesMessageService.class);
+    bind(String.class).annotatedWith(named("resourceFileLocation")).toInstance("message/messages");
```
Comment (asatklichov) [inline server/src/test/java/com/broadcom/lsp/cobol/TestModule.java:66]: "message/messages" folders ?
In our case they are not messages (JMS, Kafka, or AMQP, ... ), just raw-text, better would be: resourceBundles  or properties

File: server/src/test/java/com/broadcom/lsp/cobol/core/messages/PropertiesMessageServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+class PropertiesMessageServiceTest {
+
+  private static MessageService messageService;
+  private LocaleStore localeMock;
+
+  @BeforeEach
+  public void beforeAll() {
+    localeMock = mock(LocaleStore.class);
+    when(localeMock.getApplicationLocale()).thenReturn(new Locale("en"));
```
Comment (asatklichov) [inline server/src/test/java/com/broadcom/lsp/cobol/core/messages/PropertiesMessageServiceTest.java:36]: LocaleEnum.EN

File: server/src/test/java/com/broadcom/lsp/cobol/core/strategy/CobolErrorStrategyTest.java [hunk 10]
```diff
@@ -36,7 +39,9 @@ void noViableAltExceptionTest() {
 
     NoViableAltException error =
         new NoViableAltException(recognizer, stream, token, token, null, null);
-    CobolErrorStrategy strategy = new CobolErrorStrategy();
+    CobolErrorStrategy strategy = new CobolErrorStrategy(messageService);
+    when(messageService.getMessage(matches("ErrorStrategy.rulereportNoViableAlternative"), anyString()))
```
Comment (asatklichov) [inline server/src/test/java/com/broadcom/lsp/cobol/core/strategy/CobolErrorStrategyTest.java:43]: cameCased: rulereport -> ruleReport

**Review by asatklichov — COMMENTED** [review]
Nice job 👍  more simplified and complete solution now. 
I have some minor change requests. Beside that, generated VSIX should be tests on "Eclipse Che" with a snapshot.  Then we are READY to merge.

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java (hunk 4 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java:95]: To catch a scenario of invalid locale ( not supported locale).

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java (hunk 5 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java:109]: same case

File: server/src/test/java/com/broadcom/lsp/cobol/core/strategy/CobolErrorStrategyTest.java (hunk 10 above)
Comment (ap891843) [reply server/src/test/java/com/broadcom/lsp/cobol/core/strategy/CobolErrorStrategyTest.java:43]: nice catch, its actually a usecase

File: server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java (hunk 3 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java]: moving return to the last stmt in the method. return in finally might give a sonar issue of - "Jump statements should not occur in "finally" blocks"

Comment (asatklichov) [reply server/src/main/java/com/broadcom/lsp/cobol/core/messages/LocaleStoreImpl.java]: Of course I wrote, end-of-method.

**Review by asatklichov — APPROVED** [review]
🔢

**Pull request merged**